crates/api/src/trap.rs

```diff
@@ -1,11 +1,64 @@
+use crate::instance::Instance;
+use crate::r#ref::HostRef;
+use std::sync::{Arc, Mutex};
 use thiserror::Error;
 
+#[derive(Debug)]
+pub struct FrameInfo;
+
+impl FrameInfo {
+    pub fn instance(&self) -> *const Instance {
+        unimplemented!("FrameInfo::instance");
+    }
+
+    pub fn func_index() -> usize {
+        unimplemented!("FrameInfo::func_index");
+    }
+
+    pub fn func_offset() -> usize {
+        unimplemented!("FrameInfo::func_offset");
+    }
+
+    pub fn module_offset() -> usize {
+        unimplemented!("FrameInfo::module_offset");
+    }
+}
+
+#[derive(Debug)]
+pub struct TrapInfo {
+    message: String,
+    trace: Vec<FrameInfo>,
+}
+
+impl TrapInfo {
+    pub fn new<I: Into<String>>(message: I) -> Self {
+        Self {
+            message: message.into(),
+            trace: vec![],
+        }
+    }
+
+    /// Returns a reference the `message` stored in `Trap`.
+    pub fn message(&self) -> &str {
+        &self.message
+    }
+
+    pub fn origin(&self) -> Option<&FrameInfo> {
+        self.trace.first()
+    }
+
+    pub fn trace(&self) -> &[FrameInfo] {
+        &self.trace
+    }
+}
+
 /// A struct representing an aborted instruction execution, with a message
 /// indicating the cause.
 #[derive(Error, Debug)]
 #[error("Wasm trap: {message}")]
 pub struct Trap {
     message: String,
+    info: Arc<Mutex<HostRef<TrapInfo>>>,
```

Comment:
I think this'll want a test as well to be added in this PR that `Trap` is both `Send` and `Sync`, because currently I believe `Mutex<HostRef<T>>` is neither `Send` nor `Sync` (due to the usage of `Rc`)

Reply:
added `_assert_trap_is_sync_and_send` test, `UnsafeTrapInfo` is forced to be `Send`.